File: clients/Ruby/model/auto_attack.rb
require_relative 'entity_type'
class AutoAttack
    attr_accessor :pathfind_range
    attr_accessor :valid_targets
    def initialize(pathfind_range, valid_targets)
        @pathfind_range = pathfind_range
        @valid_targets = valid_targets
    end
    def self.read_from(stream)
        pathfind_range = stream.read_int()
        valid_targets = []
        stream.read_int().times do |_|
            valid_targets_element = stream.read_int()
            if valid_targets_element < 0 || valid_targets_element > 10
                raise "Unexpected tag value"
            end
            valid_targets.push(valid_targets_element)
        end
        AutoAttack.new(pathfind_range, valid_targets)
    end
    def write_to(stream)
        stream.write_int(@pathfind_range)
        stream.write_int(@valid_targets.length())
        @valid_targets.each do |element|
            stream.write_int(element)
        end
    end
end

File: clients/Python/model/auto_attack.py
from .entity_type import EntityType
class AutoAttack:
    def __init__(self, pathfind_range, valid_targets):
        self.pathfind_range = pathfind_range
        self.valid_targets = valid_targets
    @staticmethod
    def read_from(stream):
        pathfind_range = stream.read_int()
        valid_targets = []
        for _ in range(stream.read_int()):
            valid_targets_element = EntityType(stream.read_int())
            valid_targets.append(valid_targets_element)
        return AutoAttack(pathfind_range, valid_targets)
    def write_to(self, stream):
        stream.write_int(self.pathfind_range)
        stream.write_int(len(self.valid_targets))
        for element in self.valid_targets:
            stream.write_int(element)
    def __repr__(self):
        return "AutoAttack(" + \
            repr(self.pathfind_range) + "," + \
            repr(self.valid_targets) + \
            ")"
